beluga/include/beluga/sensor/likelihood_field_model.hpp

```diff
@@ -161,20 +164,32 @@ class LikelihoodFieldModel {
       return std::min(squared_distance, squared_max_distance);
     };
 
-    const auto to_likelihood = [amplitude =
-                                    params.z_hit / (params.sigma_hit * std::sqrt(2 * Sophus::Constants<double>::pi())),
-                                two_squared_sigma = 2 * params.sigma_hit * params.sigma_hit,
-                                offset = params.z_random / params.max_laser_distance](double squared_distance) {
-      assert(two_squared_sigma > 0.0);
-      assert(amplitude > 0.0);
+    /// Pre-computed variables
+    const double two_squared_sigma = 2 * params.sigma_hit * params.sigma_hit;
+    assert(two_squared_sigma > 0.0);
+
+    const double amplitude = params.z_hit / (params.sigma_hit * std::sqrt(2 * Sophus::Constants<double>::pi()));
+    assert(amplitude > 0.0);
+
+    const double offset = params.z_random / params.max_laser_distance;
+
+    const auto to_likelihood = [=](double squared_distance) {
       return amplitude * std::exp(-squared_distance / two_squared_sigma) + offset;
     };
 
     const auto neighborhood = [&grid](std::size_t index) { return grid.neighborhood4(index); };
 
     // determine distances to obstacles and calculate likelihood values in-place
     // to minimize memory usage when dealing with large maps
-    auto likelihood_values = nearest_obstacle_distance_map(grid.obstacle_data(), squared_distance, neighborhood);
+    auto likelihood_values = nearest_obstacle_distance_map(grid.obstacle_mask(), squared_distance, neighborhood);
```

Comment:
It end up being likelihood values after:
```cpp
std::transform(likelihood_values.begin(), likelihood_values.end(), likelihood_values.begin(), to_likelihood);
```
If I change the value to `distance_map`, then should I create a new variable called `likelihood_values` after the previous transform?

Reply:
I think that we could potentially write something like this:

```cpp
auto distance_map = nearest_obstacle_distance_map(grid.obstacle_mask(), squared_distance, neighborhood);

if (params.model_unknown_space) {
    const double inverse_max_distance = 1 / params.max_laser_distance;
    const double background_distance = -two_squared_sigma * std::log((inverse_max_distance - offset) / amplitude);

    distance_map |= beluga::actions::overlay(grid.unknown_mask(), background_distance);
}

ranges::transform(distance_map, distance_map.begin(), truncate_to_max_distance);

auto& likelihood_values = distance_map;  // re-use storage
ranges::transform(distance_map, likelihood_values.begin(), to_likelihood);

return ValueGrid2<float>{std::move(likelihood_values), grid.width(), grid.resolution()};
```

---

P.D. I'm also not a fan of defining a named lambda to use later in the code. But this might not be everyone's cup of tea.

```cpp
auto distance_map = nearest_obstacle_distance_map(grid.obstacle_mask(), squared_distance, neighborhood);

if (params.model_unknown_space) {
    const double inverse_max_distance = 1 / params.max_laser_distance;
    const double background_distance = -two_squared_sigma * std::log((inverse_max_distance - offset) / amplitude);

    distance_map |= beluga::actions::overlay(grid.unknown_mask(), background_distance);
}

ranges::transform(distance_map, distance_map.begin(),
  [squared_max_distance](double squared_distance) {
    return std::min(squared_distance, squared_max_distance);
  });

auto& likelihood_values = distance_map;  // re-use storage

ranges::transform(distance_map, likelihood_values.begin(),
  [amplitude, two_squared_sigma, offset](double squared_distance) {
    return amplitude * std::exp(-squared_distance / two_squared_sigma) + offset;
  });

return ValueGrid2<float>{std::move(likelihood_values), grid.width(), grid.resolution()};
```

Reply:
After a good night of sleep, I realized that the move-transform-assign pattern can actually be written like this:

```cpp
auto distance_map = nearest_obstacle_distance_map(grid.obstacle_mask(), squared_distance, neighborhood);

if (params.model_unknown_space) {
    const double inverse_max_distance = 1 / params.max_laser_distance;
    const double background_distance = -two_squared_sigma * std::log((inverse_max_distance - offset) / amplitude);

    distance_map |= beluga::actions::overlay(grid.unknown_mask(), background_distance);
}

auto likelihood_values = std::move(distance_map) |
                         ranges::actions::transform(truncate_to_max_distance) |
                         ranges::actions::transform(to_likelihood);

return ValueGrid2<float>{std::move(likelihood_values), grid.width(), grid.resolution()};
```

This should re-use the storage.

Reply:
Compiler is throwing an error (in testing) when trying to use `std::move()` with `|` operator. The error is due, and cite `view.hpp`:
```
            // **************************************************************************
            // *    When piping a range into an adaptor, the range must satisfy the     *
            // *    "viewable_range" concept. A range is viewable when either or both   *
            // *    of these things are true:                                           *
            // *      - The range is an lvalue (not a temporary object), OR             *
            // *      - The range is a view (not a container).                          *
            // **************************************************************************
```
Doing the following works just fine:
```cpp
auto likelihood_values = distance_map | ranges::views::transform(truncate_to_max_distance) |
                     ranges::views::transform(to_likelihood) | ranges::to<std::vector<float>>;
```
However, this won't re-use the storage.

I tried to find a workaround using `ranges::view::all(std::move(distance_map))`, but it didn't work either.

Reply:
Yeah, you should use `ranges::actions::transform` instead of `ranges::views::transform`.

You will also need to include `#include <range/v3/action/transform.hpp>`.

And finally, you can force the linter to keep the newline by adding an empty comment to the EOL in a range pipeline. Personally I think this is much better for readability.

```cpp
auto likelihood_values = distance_map |                                          //
                         ranges::actions::transform(truncate_to_max_distance) |  //
                         ranges::actions::transform(to_likelihood);
```

Reply:
> Yeah, you should use `ranges::actions::transform` instead of `ranges::views::transform`.

My bad, I didn't notice it. It's working now :tada:. I've just updated the PR.